### ISSUE TITLE
Validate no-environment should perform transform/filter VRL checks

### ISSUE THREAD

**Issue by sbalmos** [vectordotdev/vector#15037]
### A note for the community

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Use Cases

Vector validate --no-environment should perform VRL syntax checks of transforms and filters, to catch VRL errors in CI/CD-based config generation scenarios.

### Attempted Solutions

A partial workaround exists as discussed in https://github.com/vectordotdev/vector/discussions/13726. However, this only covers a hacky rewrite of sinks, not sources, and also requires the installation of the yq utility. The idea of having vector validate running against a whole Helm chart file is out of scope - this is merely a reference general idea/workaround.

### Proposal

Currently, performing a validate --no-environment does very basic validations of the provided config. Notably, it does not perform validation of any transform VRL or filter conditions. This seems to be a critical gap in CI/CD situations. For reference, I generate the Vector config from templates in Ansible, save to a temporary file, and attempt to perform a vector validate --no-environment to sanity/syntax check the file before the config contents are embedded into a Helm chart values file. The --no-environment flag performed no syntax checking, leaving an any-to-string type mismatch error in a transform until the Helm chart was deployed, causing a pod error.

--no-environment should bypass environmental checks (env vars, connecting to sources/sinks, etc), but still perform VRL validation.

### Version

0.24.2

**Comment by spencergilbert** [comment]
👋 Thanks for the issue. This duplicates https://github.com/vectordotdev/vector/issues/7198 but I think it might be more complete than the existing issue. @jszwedko which do you think we should keep open?

**Comment by jszwedko** [comment]
I closed the other as a duplicate since this issue has more information. Thanks for filing @sbalmos !

**Comment by smitthakkar96** [comment]
@jszwedko did something change, even without `--no-environment` flag we are not seeing any errors related to VRL syntax/expressions. Only workaround I can think of is to have a one test case and run `vector test` along with validate

**Comment by smitthakkar96** [comment]
When can this be prioritised? It's a blocker for us to adopt Vector in our org as we offer vector as a self-service for exclusion filtering, which means our devs independently make changes and one merge it's deployed to all clusters. Not having ability to validate VRL expressions in CI would mean if there is an invalid expression and PR is merged Vector would crash.

**Comment by jszwedko** [comment]
It should compile VRL programs when not using `--no-environment`. I just tested this again with 0.30.0 and it seems to work. If it isn't for you, could you open a bug report?

Unfortunately it still doesn't validate VRL programs when running with `--no-environment`, for now.

**Comment by smitthakkar96** [comment]
Hey @jszwedko,

You are right. My bad, without `--no-environment` it does check VRL expression

**Comment by smitthakkar96** [comment]
Some more challenges of not having VRL expression not being compiled when `--no-environment` flag is set are: 

We currently use `vector validate` in an internal k8s operator to validate the admission webhook. When two admission review requests come in parallel, `ports` and `data_dir` conflict, causing validation to fail. Example:

```
x Source \"datadog_agents\": TCP bind failed: Address already in use (os error 98)

2023-08-24T14:33:58.240518Z ERROR vector::validate: Failed to remove temporary directory. path=\"/vector-data-dir/validate_tmp\" error=No such file or directory (os error 2)
"}	{"error": "failed to validate vector configuration: {\"exit_code\":78,\"original_error\":{\"Stderr\":null},\"vector_validate_output\":\"Loaded with \\nComponent errors\\n----------------\\nx Source \\\"datadog_agents\\\": TCP bind failed: Address already in use (os error 98)\\n\\n2023-08-24T14:33:58.240518Z ERROR vector::validate: Failed to remove temporary directory. path=\\\"/vector-data-dir/validate_tmp\\\" error=No such file or directory (os error 2)\\n\"}"}\
```

Currently, we worked around this by randomizing these values during validation, but it would be nice if we didn't have to do these hacks.

I just put it out there, if it helps to make a case for prioritisation.

**Comment by yruss972** [comment]
First time vector user and this has been very unintuitive and super-annoying.
We are trying to use vector validate to catch configuration syntax issues in the docker image build.
We started with --skip-healthchecks thinking we don't need to test connectivity to sources, sinks etc. but
  1. vector still tries to find valid GCP credentials for sources and sinks, maybe tries to connect (I haven't gotten that far)
  2. vector performs unnecessary validation of tls.crt_file and tls.key_file contents which will be mounted in to the container from secret manager in prod

We switched to --no-environment not realizing that all the vrl syntax issues would go away uncaught.

Now we're stuck in the middle having to review the --skip-healthchecks output manually and choosing when to ignore the errors.

Am I missing something which can make this better?

**Comment by pront** [comment]
Hey all, 

VRL compilation is bundled into the same code path as environment-dependent operations (loading enrichment tables, instantiating sources/sinks, validating TLS certs, etc.). I think there's a simple fix for this issue. When `--no-environment` is set, after config structure validation, we should iterate over all transforms and call their `build()` method with a default context.

We don't have an ETA for this fix on our end, but we'd be happy to review a community contribution if anyone wants to pick this up.

**Comment by pront** [comment]
This was actually not that tricky to implement (with one exception): https://github.com/vectordotdev/vector/pull/25161

We will probably include this in the next-next release.
